documentation/api/query/v4/pql.markdown

```diff
@@ -0,0 +1,361 @@
+
+* single quoted - no escaping, just straight text
+* double quoted - supports escape characters
+
+Double quoted strings follow the same rules as JSON strings, and can accept escape characters:
+
+* `\n` - newline
+* `\r` - carriage return
+* `\b` - backspace
+* `\f` - formfeed
+* `\t` - tab
+* `\uXXXX` - unicode character
+
+For example to match on a string with a newline in it:
+
+    facts { value = "first line\nsecondline" }
+
+#### Booleans
+
+Booleans are represented by using the bar words: `true` or `false`.
+
+#### Numbers
+
+Numbers in PQL can be either integers or reals:
+
+    4
+    4.1
+    -10245
+    -124.012
+
+For real numbers, scientific notation is expressed using E notation:
+
+    4.1E123
+    -3.2E-123
+
+E notation follows the same rules as JSON, but currently is only
+accepted for real numbers, not integers.
+
+#### Lists
+
+Lists are groups of other literal values, and are expressed using
+brackets with elements separated by commas:
+
+    ['a', 'b', 'c', 'd']
+
+Currently lists are only supported with the `in` operator.
+
+### Implicit Subqueries
+
+Implicit subqueries works very much the same way as the `in` operator provides, however the relationship
+between some entities is clear. When an implicit relationship exists between two entity types, you can
+avoid the overhead of having to provide the join columns like with the `in` operator by using implicit
+subqueries instead.
+
+Basically, an implicit subquery looks like a query, embedded within the filter of a PQL query:
+
+    nodes {
+      facts { name = 'operatingsystem' and value = 'Debian' }
+    }
+
+In this example, while the query context is set to `nodes`, we will only return `nodes` that have
+a fact `name` of `operatingsystem` and `value` of `Debian` (so only Debian nodes basically).
+
+As mentioned this often allows you to avoid having to know which fields are required like with an `in`
+operator, but only some relationships are well defined. See the [entities] documentation for each
+entity to learn which implicit subqueries are provided automatically.
+
+Also, implicit subqueries are like any other conditional operator, so therefore can be combined with
+basic filters. The following query combines the fact subquery as before, included with a `certname`
+match on the node itself:
+
+    nodes {
+      facts { name = 'operatingsystem' and value = 'Debian' } and
+      certname ~ '^web'
+    }
+
+You can even combined with other implict subqueries, to provide more complex matching capabilities. This
+query combined everything before, but with a `resource` subquery for `Package[tomcat]`:
+
+    nodes {
+      facts { name = 'operatingsystem' and value = 'Debian' } and
+      resources { type = 'Package' and title = 'tomcat' } and
+      certname ~ '^web'
+    }
+
+## Modifiers
+
+Modifiers to the query provide extra facilities beyond a filter for limiting and changing the resultsets
+returned in special ways.
+
+### `group by`
```

Comment:
As discussed, this work and discussion has moved to here: https://tickets.puppetlabs.com/browse/PDB-2357 and the mailing list discussion